### PR TITLE
[SPARK-36136][SQL][TESTS] Refactor PruneFileSourcePartitionsSuite etc to a different package

### PR DESCRIPTION
**Pull request by sunchao**
### What changes were proposed in this pull request?

Move both `PruneFileSourcePartitionsSuite` and `PrunePartitionSuiteBase` to the package `org.apache.spark.sql.execution.datasources`. Did a few refactoring to enable this.

### Why are the changes needed?

Currently both `PruneFileSourcePartitionsSuite` and `PrunePartitionSuiteBase` are in package `org.apache.spark.sql.hive.execution` which doesn't look correct as these tests are not specific to Hive. Therefore, it's better to move them into `org.apache.spark.sql.execution.datasources`, the same place where the rule `PruneFileSourcePartitions` is at.

### Does this PR introduce _any_ user-facing change?

No, it's just test refactoring.

### How was this patch tested?

Using existing tests:
```
build/sbt "sql/testOnly *PruneFileSourcePartitionsSuite"
```
and 
```
build/sbt "hive/testOnly *PruneHiveTablePartitionsSuite"
```

### CODE REVIEW

**Comment by sunchao** [comment]
cc @viirya @dongjoon-hyun @cloud-fan @HyukjinKwon @LuciferYang . [This commit](https://github.com/apache/spark/pull/33564/commits/4663ab4098f4e439f35c804d88a36ff5d4f14ea9) fixes the test failures.

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @sunchao .

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitionsSuite.scala
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitionsSuite.scala:139]: Got it.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
I'll leave this to @viirya since he already merged this once.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitionsSuite.scala
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitionsSuite.scala:110]: why need to add `coalesce`?

Comment (huaxingao) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitionsSuite.scala:110]: Same question here. Looks good otherwise.

Comment (sunchao) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitionsSuite.scala:110]: @viirya @huaxingao this is because `TestHiveSingleton` uses `local[1]` while `SharedSparkSession` uses `local[2]` so it will generate 2 Parquet files. As result, `ParquetScan` will generate 2 input splits which makes the test fail since the provided value for `expectedPartitionCount` is 1.

**Comment by viirya** [comment]
I see. lgtm. I am not in front of laptop. Will merge this after I come back.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Thanks! Merging to master/3.2.

**Review by LuciferYang — COMMENTED** [review]
LGTM

**Pull request closed without merging**
